modules/hal_rpi_pico/CMakeLists.txt

```diff
@@ -8,60 +8,66 @@ if(CONFIG_HAS_RPI_PICO)
-		${common_dir}/pico_base/include
-		${ZEPHYR_BASE}/include
-		)
-
-	target_link_options(boot_stage2 PRIVATE
-		"-nostartfiles"
-		"--specs=nosys.specs"
-		"LINKER:--script=${boot_stage_dir}/boot_stage2.ld"
-		)
-
-	# The second stage bootloader is compiled without kconfig definitions.
-	# Therefore, in order to use toolchain.h, it needs to define CONFIG_ARM.
-	target_compile_definitions(boot_stage2 PRIVATE -DCONFIG_ARM=1)
-
-	# Generates a binary file from the compiled bootloader
-	add_custom_target(bootloader_bin ALL DEPENDS boot_stage2.bin)
-	add_custom_command(OUTPUT boot_stage2.bin
-		DEPENDS boot_stage2
-		COMMAND ${CMAKE_OBJCOPY} -Obinary $<TARGET_FILE:boot_stage2> boot_stage2.bin
-		)
-
-	# Checksums the binary, pads it, and generates an assembly file
-	add_custom_target(boot_stage2_asm ALL DEPENDS boot_stage2.S)
-	add_custom_command(OUTPUT boot_stage2.S
-		DEPENDS boot_stage2.bin
-		COMMAND ${Python3_EXECUTABLE} ${boot_stage_dir}/pad_checksum
-		-s 0xffffffff boot_stage2.bin boot_stage2.S)
+	# The Second Stage Bootloader is only linked to the app that resides
+	# at 0x100. Therefore, only if the app's offset is 0x100, the second
+	# stage bootloader should be compiled.
+	if(${CONFIG_RP2_REQUIRES_SECOND_STAGE_BOOT})
+		# Compile the second stage bootloader as a separate executable
+
+		add_executable(boot_stage2)
+
+		target_sources_ifdef(CONFIG_RP2_FLASH_W25Q080 boot_stage2 PRIVATE
+					${boot_stage_dir}/boot2_w25q080.S)
+		target_sources_ifdef(CONFIG_RP2_FLASH_GENERIC_03H boot_stage2 PRIVATE
+					${boot_stage_dir}/boot2_generic_03h.S)
+		target_sources_ifdef(CONFIG_RP2_FLASH_IS25LP080 boot_stage2 PRIVATE
+					${boot_stage_dir}/boot2_is25lp080.S)
+		target_sources_ifdef(CONFIG_RP2_FLASH_W25X10CL boot_stage2 PRIVATE
+					${boot_stage_dir}/boot2_w25x10cl.S)
+		target_sources_ifdef(CONFIG_RP2_FLASH_AT25SF128A boot_stage2 PRIVATE
+					${boot_stage_dir}/boot2_at25sf128a.S)
+
+		target_include_directories(boot_stage2 PUBLIC
+			${CMAKE_CURRENT_LIST_DIR}
+			${boot_stage_dir}/asminclude
+			${rp2_common_dir}/pico_platform/include
+			${rp2040_dir}/hardware_regs/include
+			${common_dir}/pico_base/include
+			${ZEPHYR_BASE}/include
+			)
+
+		target_link_options(boot_stage2 PRIVATE
+			"-nostartfiles"
+			"--specs=nosys.specs"
+			"LINKER:--script=${boot_stage_dir}/boot_stage2.ld"
+			)
+
+		# The second stage bootloader is compiled without kconfig definitions.
+		# Therefore, in order to use toolchain.h, it needs to define CONFIG_ARM.
+		target_compile_definitions(boot_stage2 PRIVATE -DCONFIG_ARM=1)
+
+		# Generates a binary file from the compiled bootloader
+		add_custom_target(bootloader_bin ALL DEPENDS boot_stage2.bin)
+		add_custom_command(OUTPUT boot_stage2.bin
+			DEPENDS boot_stage2
+			COMMAND ${CMAKE_OBJCOPY} -Obinary $<TARGET_FILE:boot_stage2> boot_stage2.bin
+			)
+
+		# Checksums the binary, pads it, and generates an assembly file
+		add_custom_target(boot_stage2_asm ALL DEPENDS boot_stage2.S)
```

Comment:
actually, why is there a custom target in the first place ?

Reply:
What's the alternative?

Reply:
not having a custom target.

When the output from a custom command is used by another build step, then CMake will ensure that command is invoked.

Only if multiple targets depends on the same output from a custom command, then it is required to create a custom target. But in such a case, the CMake code should have a dependency to the custom target, not to the output of the custom command.
This doesn't appear to be what you're doing, cause as far as I can tell, no targets depends on the `boot_stage2_asm` target.

See more: https://cmake.org/cmake/help/latest/command/add_custom_command.html

Reply:
@tejlmand I missed that one, I removed the `bootloader_bin` custom target successfully, but `boot_stage2_asm` seems to be needed, as the target that depends on it is in the other file. Is there any way around this, or can I keep that custom target?

Reply:
and I missed the request for changes because you moved the code around 😆 

The most correct thing to do here, is to add post build commands to the `boot_stage2` target and specify the byproducts.
 
I'm just giving you the patch set right here:
```patch
diff --git a/modules/hal_rpi_pico/bootloader/CMakeLists.txt b/modules/hal_rpi_pico/bootloader/CMakeLists.txt
index 68d080b46f..f582604169 100644
--- a/modules/hal_rpi_pico/bootloader/CMakeLists.txt
+++ b/modules/hal_rpi_pico/bootloader/CMakeLists.txt
@@ -51,15 +51,15 @@ target_link_options(boot_stage2 PRIVATE
 target_compile_definitions(boot_stage2 PRIVATE -DCONFIG_ARM=1)
 
 # Generates a binary file from the compiled bootloader
-add_custom_target(bootloader_bin DEPENDS boot_stage2.bin)
-add_custom_command(OUTPUT boot_stage2.bin
-  DEPENDS boot_stage2
+add_custom_command(TARGET boot_stage2
+  POST_BUILD
+  BYPRODUCTS boot_stage2.bin
   COMMAND ${CMAKE_OBJCOPY} -Obinary $<TARGET_FILE:boot_stage2> boot_stage2.bin
   )
 
 # Checksums the binary, pads it, and generates an assembly file
-add_custom_target(boot_stage2_asm ALL DEPENDS boot_stage2.S)
-add_custom_command(OUTPUT boot_stage2.S
-  DEPENDS boot_stage2.bin
+add_custom_command(TARGET boot_stage2
+  POST_BUILD
+  BYPRODUCTS boot_stage2.S
   COMMAND ${PYTHON_EXECUTABLE} ${boot_stage_dir}/pad_checksum
   -s 0xffffffff boot_stage2.bin boot_stage2.S)
```

that will give you all you need.

Reply:
fixed